src/lib/statistics/chunk_statistics/histograms/equal_element_count_histogram.cpp

```diff
@@ -0,0 +1,206 @@
+    DebugAssert(mins[bin_id].find_first_not_of(supported_characters) == std::string::npos, "Unsupported characters.");
+    DebugAssert(maxs[bin_id].find_first_not_of(supported_characters) == std::string::npos, "Unsupported characters.");
+    DebugAssert(mins[bin_id] <= maxs[bin_id], "Cannot have upper bin edge higher than lower bin edge.");
+
+    if (bin_id < maxs.size() - 1) {
+      DebugAssert(mins[bin_id] < mins[bin_id + 1], "Bins must be sorted and cannot overlap.");
+      DebugAssert(maxs[bin_id] < maxs[bin_id + 1], "Bins must be sorted and cannot overlap.");
+    }
+  }
+}
+
+template <typename T>
+EqualElementCountBinStats<T> EqualElementCountHistogram<T>::_get_bin_stats(
+    const std::vector<std::pair<T, HistogramCountType>>& value_counts, const BinID max_bin_count) {
+  // If there are fewer distinct values than the number of desired bins use that instead.
+  const auto bin_count = value_counts.size() < max_bin_count ? static_cast<BinID>(value_counts.size()) : max_bin_count;
+
+  // Split values evenly among bins.
+  const HistogramCountType distinct_count_per_bin = static_cast<HistogramCountType>(value_counts.size() / bin_count);
+  const BinID bin_count_with_extra_value = value_counts.size() % bin_count;
+
+  std::vector<T> mins;
+  std::vector<T> maxs;
+  std::vector<HistogramCountType> heights;
+  mins.reserve(bin_count);
+  maxs.reserve(bin_count);
+  heights.reserve(bin_count);
+
+  auto begin_index = 0ul;
+  for (BinID bin_index = 0; bin_index < bin_count; bin_index++) {
+    auto end_index = begin_index + distinct_count_per_bin - 1;
+    if (bin_index < bin_count_with_extra_value) {
+      end_index++;
+    }
+
+    mins.emplace_back(value_counts[begin_index].first);
+    maxs.emplace_back(value_counts[end_index].first);
+    heights.emplace_back(std::accumulate(
+        value_counts.cbegin() + begin_index, value_counts.cbegin() + end_index + 1, HistogramCountType{0},
+        [](HistogramCountType a, const std::pair<T, HistogramCountType>& b) { return a + b.second; }));
+
+    begin_index = end_index + 1;
+  }
+
+  return {mins, maxs, heights, distinct_count_per_bin, bin_count_with_extra_value};
+}
+
+template <typename T>
+std::shared_ptr<EqualElementCountHistogram<T>> EqualElementCountHistogram<T>::from_segment(
+    const std::shared_ptr<const BaseSegment>& segment, const BinID max_bin_count,
+    const std::optional<std::string>& supported_characters, const std::optional<uint32_t>& string_prefix_length) {
+  const auto value_counts = AbstractHistogram<T>::_value_counts_in_segment(segment);
+
+  if (value_counts.empty()) {
+    return nullptr;
+  }
+
+  const auto bin_stats = EqualElementCountHistogram<T>::_get_bin_stats(value_counts, max_bin_count);
+
+  if constexpr (std::is_same_v<T, std::string>) {
+    const auto [characters, prefix_length] =  // NOLINT (Extra space before [)
+        get_default_or_check_prefix_settings(supported_characters, string_prefix_length);
```

Comment:
So I do not forget this later: `get_default_or_check_prefix_settings` is a pretty generic name for a global function and I honestly don't know what I'd expect it to do, reading just the name

Reply:
Changed it to `get_default_or_check_string_histogram_prefix_settings`, is that better?
Frankly, I don't like this function much at all, but that code is required in quite a few places.

Reply:
If it is so internal and obscure, why not just make it a private function in the AbstractHistogram?

Reply:
Because the utils use it as well, and they are likely to be used in the wrapper for column statistics we are about to re-write.
If we do not use them there, we can move all of that to the histogram.